File: Kuroba/app/src/main/java/com/github/k1rakishou/chan/ui/controller/LoadingViewController.java
package com.github.k1rakishou.chan.ui.controller;

import static android.view.View.GONE;
import static android.view.View.VISIBLE;
import static com.github.k1rakishou.chan.utils.AppModuleAndroidUtils.getString;

import android.content.Context;
import android.widget.ProgressBar;

import com.github.k1rakishou.chan.R;
import com.github.k1rakishou.chan.core.di.component.activity.ActivityComponent;
import com.github.k1rakishou.chan.ui.theme.widget.ColorizableTextView;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import kotlin.Unit;
import kotlin.jvm.functions.Function0;

public class LoadingViewController extends BaseFloatingController {
    private ColorizableTextView loadingControllerTitle;
    private ColorizableTextView loadingControllerMessage;
    private ProgressBar progressBar;

    private String title;
    private boolean indeterminate;
    private boolean backAllowed = false;
    private @Nullable Function0<Unit> cancellationFunc;

    @Override
    protected void injectDependencies(@NotNull ActivityComponent component) {
        component.inject(this);
    }

    public LoadingViewController(Context context, boolean indeterminate) {
        super(context);

        this.indeterminate = indeterminate;
        this.title = getString(R.string.doing_heavy_lifting_please_wait);
    }

    public LoadingViewController(Context context, boolean indeterminate, String title) {
        super(context);

        this.indeterminate = indeterminate;
        this.title = title;
    }

    @Override
    public void onCreate() {
        super.onCreate();

        loadingControllerTitle = view.findViewById(R.id.loading_controller_title);
        loadingControllerTitle.setText(title);

        loadingControllerMessage = view.findViewById(R.id.loading_controller_message);
        progressBar = view.findViewById(R.id.progress_bar);
    }

    @Override
    public void onDestroy() {
        super.onDestroy();

        if (cancellationFunc != null) {
            cancellationFunc.invoke();
            cancellationFunc = null;
        }
    }

    public void enableBack() {
        backAllowed = true;
    }

    public void enableBack(@NotNull Function0<Unit> cancellationFunc) {
        backAllowed = true;
        this.cancellationFunc = cancellationFunc;
    }

    // Disable the back button for this controller unless otherwise requested by the above
    @Override
    public boolean onBack() {
        if (backAllowed) {
            presentedByController.onBack();
        }

        if (cancellationFunc != null) {
            cancellationFunc.invoke();
            cancellationFunc = null;
        }

        return true;
    }

    /**
     * Shows a progress bar with percentage in the center (cannot be used with indeterminate)
     */
    public void updateProgress(int percent) {
        if (indeterminate) {
            return;
        }

        loadingControllerMessage.setVisibility(VISIBLE);
        progressBar.setVisibility(VISIBLE);
        loadingControllerMessage.setText(String.valueOf(percent > 0 ? percent : "0"));
    }

    /**
     * Hide a progress bar and instead of percentage any text may be shown
     * (cannot be used with indeterminate)
     */
    public void updateWithText(String text) {
        if (indeterminate) {
            return;
        }

        loadingControllerMessage.setVisibility(VISIBLE);
        progressBar.setVisibility(GONE);
        loadingControllerMessage.setText(text);
    }

    @Override
    protected int getLayoutId() {
        return R.layout.controller_loading_view;
    }

}

File: Kuroba/app/src/main/java/com/github/k1rakishou/chan/ui/controller/LoadingViewController.kt
package com.github.k1rakishou.chan.ui.controller

import android.content.Context
import android.view.View
import android.widget.ProgressBar
import com.github.k1rakishou.chan.R
import com.github.k1rakishou.chan.core.di.component.activity.ActivityComponent
import com.github.k1rakishou.chan.ui.controller.base.BaseFloatingController
import com.github.k1rakishou.chan.ui.theme.widget.ColorizableBarButton
import com.github.k1rakishou.chan.ui.theme.widget.ColorizableTextView

class LoadingViewController : BaseFloatingController {
  private lateinit var loadingControllerTitle: ColorizableTextView
  private lateinit var loadingControllerMessage: ColorizableTextView
  private lateinit var cancelButton: ColorizableBarButton
  private lateinit var progressBar: ProgressBar

  private var _title: String
  private var _indeterminate: Boolean
  private var _cancelAllowed = false
  private var _cancellationFunc: Function0<Unit>? = null

  override fun injectActivityDependencies(component: ActivityComponent) {
    component.inject(this)
  }

  override fun getLayoutId(): Int {
    return R.layout.controller_loading_view
  }

  constructor(context: Context, indeterminate: Boolean) : super(context) {
    _indeterminate = indeterminate
    _title = appResources.string(R.string.doing_heavy_lifting_please_wait)
  }

  constructor(context: Context, indeterminate: Boolean, title: String) : super(context) {
    _indeterminate = indeterminate
    _title = title
  }

  override fun onCreate() {
    super.onCreate()

    loadingControllerTitle = view.findViewById<ColorizableTextView>(R.id.loading_controller_title)
    progressBar = view.findViewById<ProgressBar>(R.id.progress_bar)
    cancelButton = view.findViewById<ColorizableBarButton>(R.id.loading_controller_cancel_button)

    loadingControllerTitle.text = _title
    loadingControllerMessage = view.findViewById<ColorizableTextView>(R.id.loading_controller_message)

    if (_cancelAllowed) {
      cancelButton.visibility = View.VISIBLE
    } else {
      cancelButton.visibility = View.GONE
    }

    cancelButton.setOnClickListener {
      if (cancelButton.visibility != View.VISIBLE || !_cancelAllowed) {
        return@setOnClickListener
      }

      if (_cancellationFunc != null) {
        _cancellationFunc?.invoke()
        _cancellationFunc = null
      }

      pop()
    }
  }

  override fun onDestroy() {
    super.onDestroy()

    _cancellationFunc = null
  }

  fun enableCancellation(cancellationFunc: Function0<Unit>) {
    _cancellationFunc = cancellationFunc
    _cancelAllowed = true
  }

  // Disable the back button for this controller unless otherwise requested by the above
  override fun onBack(): Boolean {
    if (_cancelAllowed) {
      if (_cancellationFunc != null) {
        _cancellationFunc?.invoke()
        _cancellationFunc = null
      }

      pop()
    }

    return true
  }

  /**
   * Shows a progress bar with percentage in the center (cannot be used with indeterminate)
   */
  fun updateProgress(percent: Int) {
    if (_indeterminate) {
      return
    }

    loadingControllerMessage.visibility = View.VISIBLE
    progressBar.visibility = View.VISIBLE
    loadingControllerMessage.text = if (percent > 0) {
      percent.toString()
    } else {
      "0"
    }
  }

  /**
   * Hide a progress bar and instead of percentage any text may be shown
   * (cannot be used with indeterminate)
   */
  fun updateWithText(text: String?) {
    if (_indeterminate) {
      return
    }

    loadingControllerMessage.visibility = View.VISIBLE
    progressBar.visibility = View.GONE
    loadingControllerMessage.text = text
  }
}
